### DISCUSSION TITLE
Type variable scoping with variadic generics

### DISCUSSION THREAD

**Discussion by sylee957** [microsoft/pyright#4918] [Q&A]
I also found an issue with #4852 with variadic generics

Given the functions

```python3
from collections.abc import Callable
from typing import TypeVar, TypeVarTuple

X_0 = TypeVar("X_0")
X_1 = TypeVar("X_1")
Xs = TypeVarTuple("Xs")
Ys = TypeVarTuple("Ys")
X = TypeVar("X")
Y = TypeVar("Y")


def parallel(
    f: Callable[[X], Y],
    g: Callable[[*Xs], tuple[*Ys]],
) -> Callable[[X, *Xs], tuple[Y, *Ys]]:
    def wrapped(a: X, *bs: *Xs):
        return f(a), *g(*bs)

    return wrapped


def identity(x: X) -> X:
    return x


def parallel_identity(*xs: *Xs) -> tuple[*Xs]:
    return xs
```

This does not pass the type check

```python3
def test_1(a: X_0, b: X_1, *others: *Xs) -> tuple[X_0, X_1, *Xs]:
    return parallel(identity, parallel(identity, parallel_identity))(a, b, *others)
```

While the same trick with duplicate functions works

```python3
def identity_1(x: X) -> X:
    return x


def identity_2(x: X) -> X:
    return x


def test_2(a: X_0, b: X_1, *others: *Xs) -> tuple[X_0, X_1, *Xs]:
    return parallel(identity_1, parallel(identity_2, parallel_identity))(a, b, *others)
```

**Comment by erictraut** [comment]
This is a much more difficult case to handle. I agree that it's technically a bug, but I'm going to ignore it for now. You're combining very esoteric features of the type system that almost no one is using today, even individually. No other Python type checkers support this case or the other cases you reported recently. I might revisit this in the future if there is demand from other pyright users.

Out of curiosity, are you using this functionality in real code, or are you just playing around with type checking and trying to discover edge cases that don't work?

**Reply by sylee957** [reply]
My usage comes from developing the step-by-step math calculator, 
where I use type checker to verify the mathematical problem solving system with functional programming in Python.
I have to assist with type system to verify trigonometric identities or symbolic derivatives, as much as I can.

**Reply by erictraut** [reply]
Nice! You appear to be pushing the type system to its limits. It's good to see someone using all of these features. :)
